### ISSUE TITLE
Icon Tree Shaking not working

### ISSUE THREAD

**Issue by seanemmer** [segmentio/evergreen#790]
Version: 4.26
Bundler: Create-React-App default (Webpack)

Hello, I updated to the latest version to take advantage of tree-shaking but am still seeing all of the icons in my bundle:

<img width="521" alt="image" src="https://user-images.githubusercontent.com/8183188/79790895-4ee41d80-8301-11ea-80db-ce85950bc984.png">

I am not using the `Icon` component anywhere in my app.  I do, however, use `Menu` and `Menu.Item` and am wondering if the icon prop on Menu.Item is causing all icons to be included.

**Comment by mshwery** [comment]
Yeah – I believe you will not benefit from tree-shaking in v4.x if you use any components that internally rely on `Icon` (e.g. `Menu.Item`, `IconButton`, maybe even `Button`?)

**Comment by mshwery** [comment]
To get icon tree-shaking across all of evergreen we'll need to implement a breaking change that will require rewriting all imports and usage of icons to pass explicit icon components, instead of using icon strings like `icon="cog"`

**Comment by seanemmer** [comment]
Interesting.  An alternative approach (what Fontawesome does) is to allow users to define a custom icon library, and only allow icon strings for icons that have been added to the library:

https://fontawesome.com/how-to-use/with-the-api/setup/library

**Comment by mshwery** [comment]
Interesting approach. So while still supporting icon strings, you'd have to preload them.

Instead of reading from a map of all possible strings, we would be reading from the map of icons explicitly imported. That's a good idea. Would require some additional setup – and could still be considered a breaking change since you can't safely upgrade without adding that registration step.

**Comment by cstrat** [comment]
I use FontAwesome in my app and I don't use strings but rather import the icons and reference them as components:

```
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {  faHome } from '@fortawesome/pro-solid-svg-icons';

<FontAwesomeIcon icon={faHome} color="red" size="sm" />
```

I would love to be able to use these in evergreen somehow like this:

`<Button iconBefore={faHome}>Go Home</Button>`

**Comment by mshwery** [comment]
@cstrat you can do that today, but you'd have to pass the entire component: `iconBefore={<FontAwesomeIcon icon={faHome} />}`

**Comment by cstrat** [comment]
Game changer, I didn't realise that would work!

Just playing with it now, the only issue is I can't add the margin to my `<FontAwesome>` component without doing something hacky (adding a className with some hardcoded CSS to add spacing).

It would be nice if Evergreen added the margin to the injected icon. I might look into the code to see how that might work.

**Comment by cstrat** [comment]
So Evergreen does add the correct margins etc... however it looks like the `<Icon>` component ignores all props if you pass a node into it rather than use one of the Blueprint Icons.
I don't understand the reasoning. @mshwery do you know if this is just something pulled in from the blueprint codebase?

https://github.com/segmentio/evergreen/blob/master/src/icon/src/Icon.js#L31-L33

**Comment by mshwery** [comment]
@cstrat yeah, that comment looks based on blueprintjs. I think you could do something like this:

```js
const Icon = ({ icon, ...props }) => <Box is={icon} {...props} />

// later...
return <Button iconBefore={<Icon icon={faHome} marginRight={8} />}>Hello</Button>
```

**Comment by seanemmer** [comment]
It doesn't necessarily have to be a breaking change - FontAwesome's setup is opt-in.  It defaults to loading the entire icon set, but if you call library.add, then it will only load the icons you specify in your library.

**Comment by totaldis** [comment]
Just for my understanding.

If you were using the `<Icon/>` component before `4.26.0` it was bundling all icons from blueprintjs, but now it's only bundling those you've specifically asked for. 

But if you use a component that internally relies on `<Icon />` then it still bundles all icons from blueprintjs? It actually looks more like the icons themselves were created as separate components and the svg content was pulled in. So now the individual icons are addressable, but there's some wrapping to allow the string "icon" to still work.

If that's all true then at least I'm grasping the situation. I too am in @seanemmer's situation, but previously I was swapping blueprints icons out with `NormalModuleReplacementPlugin` which was simple since their exports all had one of two names: `IconSvgPaths16` and `IconSvgPaths20`. 

In this latest version I'd have to accomplish that by replacing every icon component with some generic component (except for those that are necessary since they're baked into some components like <CrossIcon />) . 

In other words, version `4.25.0`:
![image](https://user-images.githubusercontent.com/3299536/80144313-44818800-857c-11ea-9f01-ab9e93cda251.png)

version `4.26.0`:
![image](https://user-images.githubusercontent.com/3299536/80144365-5a8f4880-857c-11ea-91c8-3a16549667cc.png)

Again, I'm not looking for action here, just want to know if my assumptions are wrong which may alater my approach. Thanks. 😉

**Comment by mshwery** [comment]
Hi @totaldis! Yes there are some incorrect assumptions in there 😅 

>If you were using the <Icon/> component before 4.26.0 it was bundling all icons from blueprintjs, but now it's only bundling those you've specifically asked for.

`<Icon />` behavior is unchanged. If you use this component you will not get any tree-shaking. Similarly if you use any other Evergreen component that internally uses `<Icon />` you will not get the benefit of tree-shaking (plans to fix this as a breaking change in v5).

>It actually looks more like the icons themselves were created as separate components

Correct. If you want to tree-shake icons you should import them individually (e.g. `import { CogIcon } from 'evergreen-ui'`. There is a codemod in the package under `/codemods/` that will automatically do this for you if you want to use it.

>...but previously I was swapping blueprints icons out with `NormalModuleReplacementPlugin` which was simple since their exports all had one of two names: `IconSvgPaths16` and `IconSvgPaths20`.
> 
> In this latest version I'd have to accomplish that by replacing every icon component with some generic component (except for those that are necessary since they're baked into some components like `<CrossIcon />`) .

Correct.

**Comment by Chrischuck** [comment]
@mshwery is this something we'd want to add to v5? since it is a breaking change and all or are looking for a solution that's backwards compatible

**Comment by mshwery** [comment]
> @mshwery is this something we'd want to add to v5? since it is a breaking change and all or are looking for a solution that's backwards compatible

@Chrischuck 100% will be part of v5: 
<img width="782" alt="Screen Shot 2020-04-28 at 1 16 11 PM" src="https://user-images.githubusercontent.com/710752/80522601-73b84080-8952-11ea-93af-d1baaa20724b.png">

**Comment by mshwery** [comment]
https://github.com/segmentio/evergreen/pull/818

**Comment by mshwery** [comment]
@totaldis @cstrat @seanemmer @Chrischuck if you want to take v5 for a spin (WIP / unstable) you can try out tree shaking there `yarn add evergreen-ui@next`

Notably version `5.0.0-2` should not only allow for complete tree-shaking of icons, but also it fully supports custom icon components.

**Comment by cstrat** [comment]
Oh that sounds great, I will give it a try later this week and report back any issues.

**Comment by Chrischuck** [comment]
@mshwery I'll give it a run!

**Comment by cstrat** [comment]
@mshwery haven't done anything in depth yet - but just tested on an `<ListItem>` and it worked flawlessly!

Going to try on a few other elements.

**Comment by seanemmer** [comment]
@mshwery stoked this is up!

I was able to get it working with a simple widget that uses 4 Evergreen components but not a larger application.  Sourcemap is below, I'm guessing some components are still internally using blueprint icons (Switch, Checkbox, Toaster, Alert, Combobox, Select, Spinner)?

<img width="1781" alt="image" src="https://user-images.githubusercontent.com/8183188/83069759-4cbb6000-a01f-11ea-84e3-8cbf19d4b9b6.png">

**Comment by mshwery** [comment]
EG is using some internally, but only a few!

**Comment by mshwery** [comment]
Can you share the project code or config used to bundle/tree shake?

**Comment by seanemmer** [comment]
I'm using CRA ("react-scripts": "3.3.0").

I used the same setup for the widget app that tree-shakes correctly.

**Comment by mshwery** [comment]
I'm wondering if you were to import them via their full path TSDX would be smart enough to tree-shake them?

e.g. `import { AddIcon } from 'evergreen-ui/esm/icons/generated/AddIcon'`

**Comment by seanemmer** [comment]
Unfortunately I’m not using icons anywhere in the application so don’t
think that will help.  They are getting added as a dependency because of my
use of the general purpose components.

**Comment by seanemmer** [comment]
This was an issue with TSDX - their modules default to commonjs vs ESM, which I had to change in the configuration.

Evergreen tree shaking now working across CRA, TSDX, and Gatsby.  Thanks for implementing this, reduced our total package size by 16%!

-Sean
